modules/python/python-value-pairs.c

```diff
@@ -27,81 +27,32 @@
 #include "python-types.h"
 #include "messages.h"
 
-/** Value pairs **/
-
-static void
-add_long_to_dict(PyObject *dict, const gchar *name, long num)
-{
-  PyObject *pyobject_to_add = py_long_from_long(num);
-  if (!pyobject_to_add)
-    return;
-
-  PyDict_SetItemString(dict, name, pyobject_to_add);
-  Py_DECREF(pyobject_to_add);
-}
-
-static void
-add_string_to_dict(PyObject *dict, const gchar *name, const char *value, gsize value_len)
-{
-  PyObject *pyobject_to_add = PyBytes_FromStringAndSize(value, value_len);
-  if (!pyobject_to_add)
-    {
-      gchar buf[256];
-      _py_format_exception_text(buf, sizeof(buf));
-
-      msg_error("Error while constructing python object",
-                evt_tag_str("exception", buf));
-      _py_finish_exception_handling();
-      return;
-    }
-
-  PyDict_SetItemString(dict, name, pyobject_to_add);
-  Py_DECREF(pyobject_to_add);
-}
 
+/** Value pairs **/
 static gboolean
 python_worker_vp_add_one(const gchar *name,
                          LogMessageValueType type, const gchar *value, gsize value_len,
                          gpointer user_data)
 {
   const LogTemplateOptions *template_options = (const LogTemplateOptions *)((gpointer *)user_data)[0];
   PyObject *dict = (PyObject *)((gpointer *)user_data)[1];
-  gboolean need_drop = FALSE;
   gboolean fallback = template_options->on_error & ON_ERROR_FALLBACK_TO_STRING;
 
-  switch (type)
+  PyObject *obj = py_obj_from_log_msg_value(value, value_len, type);
```

Comment:
This is not strictly connected to typing, but what kind of feature does value-pairs provides for a python destination ?
Since the `LogMessage` is exposed to the destination it can access the macros, also template evaluation can be accessed as well.

I think the value-pairs at least from destination point of view is no longer needed, and it is even worse than just passing in the log message.

Reply:
Value-pairs is used, when you manually set value pairs to the python destination in the config. When it is set, the python code will not receive a LogMessage object, but a simple dict.

I do not like this behavior either.

Reply:
The value pairs should be removed, it should work before 4.0 and in 4.0 I would remove as is.
It is confusing and likely a performance bottleneck (not measured).

Reply:
Talked about this in a handover not so long ago, and we decided to keep it as it is.